# Google Colaboratory




---



## Google Colab 사양


* 플랫폼

In [ ]:
import platform
platform.platform()

'Linux-5.4.188+-x86_64-with-Ubuntu-18.04-bionic'

리눅스 운영체제를 사용한다

* 운영체제

In [ ]:
!cat /etc/issue.net

Ubuntu 18.04.5 LTS


Ubuntu 18.04.5 LTS 사양으르 사용하고 있음

* CPU 사양

In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0x1
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 b

CPU: 제뉴인인텔

* 메모리 사양

In [ ]:
!cat /proc/meminfo

MemTotal:       13298580 kB
MemFree:        10592208 kB
MemAvailable:   12422676 kB
Buffers:          134124 kB
Cached:          1845348 kB
SwapCached:            0 kB
Active:          1013068 kB
Inactive:        1451248 kB
Active(anon):     435792 kB
Inactive(anon):      504 kB
Active(file):     577276 kB
Inactive(file):  1450744 kB
Unevictable:           0 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:               872 kB
Writeback:             0 kB
AnonPages:        484844 kB
Mapped:           228416 kB
Shmem:              1208 kB
KReclaimable:      87796 kB
Slab:             136908 kB
SReclaimable:      87796 kB
SUnreclaim:        49112 kB
KernelStack:        4976 kB
PageTables:         7136 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:     6649288 kB
Committed_AS:    2982952 kB
VmallocTotal:   34359738367 kB
VmallocUsed:       45600 kB
VmallocChunk:          0 kB
Percpu:          

메모리 공간 확인

* 디스크 사양

In [ ]:
!df -h

Filesystem      Size  Used Avail Use% Mounted on
overlay         167G   39G  129G  24% /
tmpfs            64M     0   64M   0% /dev
shm             5.7G     0  5.7G   0% /dev/shm
/dev/root       2.0G  1.2G  812M  59% /sbin/docker-init
tmpfs           6.4G   36K  6.4G   1% /var/colab
/dev/sda1       174G   42G  132G  25% /opt/bin/.nvidia
tmpfs           6.4G     0  6.4G   0% /proc/acpi
tmpfs           6.4G     0  6.4G   0% /proc/scsi
tmpfs           6.4G     0  6.4G   0% /sys/firmware


디스크 용량이 얼마나 되는지. 대략 167기가바이트

* 파이썬 버전

In [ ]:
!python --version

Python 3.7.13


파이썬버전 3.7.13

## Google Colab 런타임
* Colab에서 고성능 하드웨어로 GPU나 TPU 사용 가능
* 런타임 유형 변경(Change runtime type) 필요
  + None: CPU만 사용
  + GPU: 하드웨어 가속으로 GPU 사용
  + TPU: 하드웨어 가속으로 TPU 사용

In [ ]:
!nvidia-smi

Sun Jul  3 13:29:55 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   38C    P0    27W / 250W |      0MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

## 파일 저장 및 다운로드/업로드

* Jupyter Notebook 환경에서 파일 저장 및 다운로드

In [ ]:
%%writefile test.txt
text

Writing test.txt


In [ ]:
! cat test.txt

text

구글 코랩에 있는 이 파일을 내 컴퓨터에 받자

In [ ]:
from google.colab import files
files.download('test.txt')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

다운로드가 된다.

In [ ]:
upload = files.upload()

Saving 1.png to 1.png


업로드도 가능.<br>
바탕화면에 있는 파일을 업로드 해봤다.

In [ ]:
!ls

1.png  sample_data  test.txt


업로드 되어있는 파일들을 확인할수 있다.

## Google Drive 연동

* Google Colab은 Google Drive와 `mount`를 통해 쉽게 연동 가능
* Google Drive에 소스 코드 저장 뿐만 아니라 파일을 열거나 저장 가능

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!ls /content/drive

MyDrive




---



# Jupyter Notebook
IPython Shell의 브라우저 기반 그래픽 인터페이스<br>
파이썬과 Python 문장 실행<br>
서식 있는 텍스트와 정적/동적 시각화, 수학 공식 표현

## help()

* 파이썬 객체에 대한 요약 정보와 사용법 보기

In [ ]:
help(min)
#min function 쓰는 방법을 알려줌

Help on built-in function min in module builtins:

min(...)
    min(iterable, *[, default=obj, key=func]) -> value
    min(arg1, arg2, *args, *[, key=func]) -> value
    
    With a single iterable argument, return its smallest item. The
    default keyword-only argument specifies an object to return if
    the provided iterable is empty.
    With two or more arguments, return the smallest argument.



## ?로 설명 보기

* 객체 요약 정보 및 사용법이 있는 docstring 보기

In [ ]:
max?
# ? docstring: 도큐먼트를 보여줌. 물음표만 넣으면 됨

In [ ]:
li = ['One', 'Two', 'Three']
li?
#내가 만든 리스트의 값과 타입과 렝쓰 등을 보여줌

In [ ]:
li.count?

In [ ]:
def power(b, n):
  """b의 n승을 반환"""
  return pow(b,n)

In [ ]:
power?
# #함수를 정의할때 첫줄의 더블 쿼테이션을 세개 연달아 해서 스트링을 해주면 추후 헬프에 연동되어 설명이 나옴

In [ ]:
power??
#접근 가능한 소스까지도 표현해줌

## 탭(tab) 자동완성

* 객체와 모듈, 인터페이스의 내용을 자동 완성

In [ ]:
li = ['One', 'Two', 'Three']
li.append
li.clear
li.count
# .을 누르면 넣을수 있는 함수가 주르륵 나온다. tab키로 빠르게 완성할수있음

##와일드카드(wildcard) 매칭

* 단어의 중간이나 마지막 글자로 매칭하여 객체나 속성 찾기

별표문양임

In [ ]:
*Error? #에러로 끝나는 모든걸 보여줌

In [ ]:
str.*index*?
#스트링 자료형에 앞뒤로 와일드카드를 붙인다 -> index 어쩌고가 있었는데..를 알고 싶을때 씀

## 매직 명령어(magic commands)

* Jupyter Notebook 환경에서 파이썬 파일 저장 및 실행

In [ ]:
%%writefile test.py 
print('Hello Colab')

Overwriting test.py


롸이트파일이 매직커맨드임. 파이썬파일을 올려보자. 주피터 노트북은 파이썬 파일이 아님


In [ ]:
%run test.py

Hello Colab


런 매직 메소드로 테스트.py 파일을 실행시킴

* 코드 실행 시간 측정

In [ ]:
%%timeit?

파이썬 문장의 표현의 시간측정할때 사용하는 매직메서드

In [ ]:
%timeit li = [n ** n for n in range(10000)]

1 loop, best of 5: 5.5 s per loop


In [ ]:
%%timeit
li = []
for n in range(10000):
  li.append(n ** n)

1 loop, best of 5: 5.49 s per loop


퍼센트 두개는 여러줄로 하는거

## 입력과 출력 이력

* In과 Out 객체를 통해 이전 명령어와 명령어의 출력 결과 보기

입력이 곧 번호 [1], [2] 처럼 나오는거. 거기에 따라 나오는 결과물이 있는것. 계속 히스토리로 유지돼서 이력이 쌓임

In [ ]:
In #지금까지 내가 명령을 내렸던 모든게 다 저장이 되어있음

['',
 "from google.colab import drive\ndrive.mount('/content/drive')",
 "get_ipython().system('ls /content/drive')",
 'help(min)\n#min function 쓰는 방법을 알려줌',
 "get_ipython().magic('pinfo max')",
 "li = ['One', 'Two', 'Three']\nget_ipython().magic('pinfo li')",
 "li = ['One', 'Two', 'Three']\nget_ipython().magic('pinfo li')\n#내가 만든 리스트의 값과 타입과 렝쓰 등을 보여줌",
 "get_ipython().magic('pinfo li.count')",
 'def power(b, n):\n  """b의 n승을 반환"""\n  return pow(b,n)',
 "get_ipython().magic('pinfo power')",
 "get_ipython().magic('pinfo power')\n#함수를 정의할때 첫줄의 더블 쿼테이션을 세개 연달아 해서 스트링을 해주면 추후 헬프에 연동되어 설명이 나옴",
 "get_ipython().magic('pinfo2 power')",
 "get_ipython().magic('pinfo2 power')\n#접근 가능한 소스까지도 표현해줌",
 "get_ipython().magic('psearch *Error')",
 'str.*index*? #스트링에 앞뒤로 와일드카드를 붙인다 -> index 어쩌고가 있었는데..를 알고 싶을때 씀',
 'str.*index*? \n#스트링에 앞뒤로 와일드카드를 붙인다 -> index 어쩌고가 있었는데..를 알고 싶을때 씀',
 "get_ipython().magic('psearch *Error')",
 'str.*index*? \n#스트링 자료형에 앞뒤로 와일드카드를 붙인다 -> index 어쩌고가 있었는데..를 알고 싶을때 씀',
 'st

In [ ]:
In[26] #26번째로 실행했던 매직 메써드가 뜸

"get_ipython().magic('pinfo %%timeit')"

In [ ]:
Out

{29: ['',
  "from google.colab import drive\ndrive.mount('/content/drive')",
  "get_ipython().system('ls /content/drive')",
  'help(min)\n#min function 쓰는 방법을 알려줌',
  "get_ipython().magic('pinfo max')",
  "li = ['One', 'Two', 'Three']\nget_ipython().magic('pinfo li')",
  "li = ['One', 'Two', 'Three']\nget_ipython().magic('pinfo li')\n#내가 만든 리스트의 값과 타입과 렝쓰 등을 보여줌",
  "get_ipython().magic('pinfo li.count')",
  'def power(b, n):\n  """b의 n승을 반환"""\n  return pow(b,n)',
  "get_ipython().magic('pinfo power')",
  "get_ipython().magic('pinfo power')\n#함수를 정의할때 첫줄의 더블 쿼테이션을 세개 연달아 해서 스트링을 해주면 추후 헬프에 연동되어 설명이 나옴",
  "get_ipython().magic('pinfo2 power')",
  "get_ipython().magic('pinfo2 power')\n#접근 가능한 소스까지도 표현해줌",
  "get_ipython().magic('psearch *Error')",
  'str.*index*? #스트링에 앞뒤로 와일드카드를 붙인다 -> index 어쩌고가 있었는데..를 알고 싶을때 씀',
  'str.*index*? \n#스트링에 앞뒤로 와일드카드를 붙인다 -> index 어쩌고가 있었는데..를 알고 싶을때 씀',
  "get_ipython().magic('psearch *Error')",
  'str.*index*? \n#스트링 자료형에 앞뒤로 와일드카드를 붙인다 -> index 어쩌고가 있

In [ ]:
Out[31]

"get_ipython().magic('pinfo %%timeit')"

In [ ]:
print("In[31:] " + In[31] + "\nOut[31]: " + Out[31])

In[31:] In[26]
Out[31]: get_ipython().magic('pinfo %%timeit')


* `_`를 이용해 이전 출력값 출력하기

In [ ]:
print(_) #언더스코어 1개 이전 출력결과

get_ipython().magic('pinfo %%timeit')


In [ ]:
print(__) #언더스코어 2개 이전 이전 출력결과

get_ipython().magic('pinfo %%timeit')


In [ ]:
print(___) #언더스코어 3개 이전 이전 이전의 출력결과

['', "from google.colab import drive\ndrive.mount('/content/drive')", "get_ipython().system('ls /content/drive')", 'help(min)\n#min function 쓰는 방법을 알려줌', "get_ipython().magic('pinfo max')", "li = ['One', 'Two', 'Three']\nget_ipython().magic('pinfo li')", "li = ['One', 'Two', 'Three']\nget_ipython().magic('pinfo li')\n#내가 만든 리스트의 값과 타입과 렝쓰 등을 보여줌", "get_ipython().magic('pinfo li.count')", 'def power(b, n):\n  """b의 n승을 반환"""\n  return pow(b,n)', "get_ipython().magic('pinfo power')", "get_ipython().magic('pinfo power')\n#함수를 정의할때 첫줄의 더블 쿼테이션을 세개 연달아 해서 스트링을 해주면 추후 헬프에 연동되어 설명이 나옴", "get_ipython().magic('pinfo2 power')", "get_ipython().magic('pinfo2 power')\n#접근 가능한 소스까지도 표현해줌", "get_ipython().magic('psearch *Error')", 'str.*index*? #스트링에 앞뒤로 와일드카드를 붙인다 -> index 어쩌고가 있었는데..를 알고 싶을때 씀', 'str.*index*? \n#스트링에 앞뒤로 와일드카드를 붙인다 -> index 어쩌고가 있었는데..를 알고 싶을때 씀', "get_ipython().magic('psearch *Error')", 'str.*index*? \n#스트링 자료형에 앞뒤로 와일드카드를 붙인다 -> index 어쩌고가 있었는데..를 알고 싶을때 씀', 'str.*index? \n#스트링 자

In [ ]:
_30  #30번째 출력해줘. 이력에 있는 수를 입력하면 됨

['',
 "from google.colab import drive\ndrive.mount('/content/drive')",
 "get_ipython().system('ls /content/drive')",
 'help(min)\n#min function 쓰는 방법을 알려줌',
 "get_ipython().magic('pinfo max')",
 "li = ['One', 'Two', 'Three']\nget_ipython().magic('pinfo li')",
 "li = ['One', 'Two', 'Three']\nget_ipython().magic('pinfo li')\n#내가 만든 리스트의 값과 타입과 렝쓰 등을 보여줌",
 "get_ipython().magic('pinfo li.count')",
 'def power(b, n):\n  """b의 n승을 반환"""\n  return pow(b,n)',
 "get_ipython().magic('pinfo power')",
 "get_ipython().magic('pinfo power')\n#함수를 정의할때 첫줄의 더블 쿼테이션을 세개 연달아 해서 스트링을 해주면 추후 헬프에 연동되어 설명이 나옴",
 "get_ipython().magic('pinfo2 power')",
 "get_ipython().magic('pinfo2 power')\n#접근 가능한 소스까지도 표현해줌",
 "get_ipython().magic('psearch *Error')",
 'str.*index*? #스트링에 앞뒤로 와일드카드를 붙인다 -> index 어쩌고가 있었는데..를 알고 싶을때 씀',
 'str.*index*? \n#스트링에 앞뒤로 와일드카드를 붙인다 -> index 어쩌고가 있었는데..를 알고 싶을때 씀',
 "get_ipython().magic('psearch *Error')",
 'str.*index*? \n#스트링 자료형에 앞뒤로 와일드카드를 붙인다 -> index 어쩌고가 있었는데..를 알고 싶을때 씀',
 'st

* `%history`를 이용한 입력 이력 살펴보기

In [ ]:
%history -n 1-7 #히스토리 내용 1번부터 7번까지 보여줘

   1:
from google.colab import drive
drive.mount('/content/drive')
   2: !ls /content/drive
   3:
help(min)
#min function 쓰는 방법을 알려줌
   4: max?
   5:
li = ['One', 'Two', 'Three']
li?
   6:
li = ['One', 'Two', 'Three']
li?
#내가 만든 리스트의 값과 타입과 렝쓰 등을 보여줌
   7: li.count?


* `%rerun`을 이용해 이전 입력 이력 다시 실행

In [ ]:
%rerun

=== Executing: ===
%history -n 1-7
=== Output: ===
   1:
from google.colab import drive
drive.mount('/content/drive')
   2: !ls /content/drive
   3:
help(min)
#min function 쓰는 방법을 알려줌
   4: max?
   5:
li = ['One', 'Two', 'Three']
li?
   6:
li = ['One', 'Two', 'Three']
li?
#내가 만든 리스트의 값과 타입과 렝쓰 등을 보여줌
   7: li.count?


## 셸 명령어

* 텍스트 기반의 셸 명령어 처리
* `!` 문자를 명령어 앞에 붙여서 셸 명령어 사용 가능

* `ls`: 디렉토리 리스트 보기

In [ ]:
!ls #이 디렉토리에 리스트가 뭐가 있는지

drive  sample_data  test.py


* `pwd`: 현재 경로 보기

In [ ]:
!pwd # 현재 경로가 어딘지

/content


* `cd`: 디렉토리 변경
* IPython에서는 임시 셸에서 실행

In [ ]:
!cd sample_data/ && ls # 샘플데이터 디렉토리 안에 있는 디렉과 파일들을 보여줌

anscombe.json		      mnist_test.csv
california_housing_test.csv   mnist_train_small.csv
california_housing_train.csv  README.md


* `%cd`: 지속적인 디렉토리 변경

In [ ]:
#이 쉘에서만 바뀌는게 아니라 지속적으로 위치가 바꼈음 좋겠다.
%cd sample_data/
#앞으로 여기로 기준이 바뀜

/content/sample_data


* `echo`: 화면 출력

In [ ]:
!echo "Shell"

Shell


* `mkdir`: 디렉토리 생성

In [ ]:
!mkdir tmp

In [ ]:
!ls #tmp 파일 생성됨

anscombe.json		      mnist_test.csv	     tmp
california_housing_test.csv   mnist_train_small.csv
california_housing_train.csv  README.md


* `cat`: 파일 보기

In [ ]:
!cat README.md #파일의 내용 출력해서 볼수있음

This directory includes a few sample datasets to get you started.

*   `california_housing_data*.csv` is California housing data from the 1990 US
    Census; more information is available at:
    https://developers.google.com/machine-learning/crash-course/california-housing-data-description

*   `mnist_*.csv` is a small sample of the
    [MNIST database](https://en.wikipedia.org/wiki/MNIST_database), which is
    described at: http://yann.lecun.com/exdb/mnist/

*   `anscombe.json` contains a copy of
    [Anscombe's quartet](https://en.wikipedia.org/wiki/Anscombe%27s_quartet); it
    was originally described in

    Anscombe, F. J. (1973). 'Graphs in Statistical Analysis'. American
    Statistician. 27 (1): 17-21. JSTOR 2682899.

    and our copy was prepared by the
    [vega_datasets library](https://github.com/altair-viz/vega_datasets/blob/4f67bdaad10f45e3549984e17e1b3088c731503d/vega_datasets/_data/anscombe.json).


* `cp`: 디렉토리/파일 복사

In [ ]:
#cp=copy
!cp README.md tmp/

In [ ]:
!ls tmp/

README.md


* `rm`: 디렉토리/파일 삭제

In [ ]:
!rm -r tmp

In [ ]:
!ls

anscombe.json		      mnist_test.csv
california_housing_test.csv   mnist_train_small.csv
california_housing_train.csv  README.md




---



# 마크다운(Markdown)

* 문법이 간단하고, 사용이 쉬움



## 제목(Heading) 표시

# 헤딩 1
## 헤딩 2
### 헤딩 3
#### 헤딩 4
##### 헤딩 5
###### 헤딩 6

## 목록(List) 구성



1. 첫째
2. 둘째
3. 셋째
* 1단계
  + 2단계 # 별표 +, - 로 심볼무늬 활용
    - 3단계

## 폰트 스타일(Font Style)



**bold**
__bold__
*italic*
_italic_
~~strike~~
<u>underscore<u>

## 인용구



> 인용 1
>> 인용 2
>>> 인용 3

## 링크



[Google](https://www.google.com)

## 이미지


## 수평선




---

## 표




| C1 | C2 | C3  |
|:---|---:|:---:|
| D1 | D2 | D3 |

작대기는 시프트 위에 파이프라인 기호 |

## 코드






`inline code`

```
block code
```

```python
def powr(b, n)
  return pow(b,n)
```


코드를 표현해 줄때 사용하는 꺽쇠 <> 또는 esc 바로 밑의 ```


# 새 섹션
**굵은 텍스트**
*이탤릭체 텍스트*


```
# 코드로 형식 지정됨
```
[링크 텍스트](https://)


> 들여쓴 블록



1.   항목 추가
2.   항목 추가

*   항목 추가
*   항목 추가

---
\
: